io-engine/src/bdev/dev.rs

```diff
@@ -66,7 +66,12 @@ pub(crate) mod uri {
             }
             "malloc" => Ok(Box::new(malloc::Malloc::try_from(&url)?)),
             "null" => Ok(Box::new(null_bdev::Null::try_from(&url)?)),
-            "nvmf" => Ok(Box::new(nvmx::NvmfDeviceTemplate::try_from(&url)?)),
+            // keeping nvmf scheme so existing tests(if any, setting this
+            // scheme) work. The replicas and nexus however should
+            // always be exposing nvmf+tcp or nvmf+rdma now.
+            "nvmf" | "nvmf+tcp" | "nvmf+rdma" => {
```

Comment:
So nvmf is now equivalent to nvmf+tcp?
What about when we expose both tcp and rdma?

Reply:
> So nvmf is now equivalent to nvmf+tcp?

Yeah, but not expecting `nvmf` to be generated anymore.

>  What about when we expose both tcp and rdma?

`nvmf` may well be removed later. Also, I will hold to merge this change until control plane changes are also ready to handle new uris.

Reply:
> nvmf may well be removed later. Also, I will hold to merge this change until control plane changes are also ready to handle new uris.

This would be a problem for back compatibility. Maybe if RDMA is not enabled we can keep using just nvmf?

Also I don't think I got the answer to, what if rdma is enabled, meaning we expose both tcp and rdma, which uri would we expose?
nvmf+rdma?
nvmf+tcp+rdma?
A list of URIs? nvmf+tcp:://....., nvmf+rdma:///....

Reply:
When RDMA is enabled, the uri will be `nvmf+rdma` (tcp will be implied).

> Maybe if RDMA is not enabled we can keep using just nvmf?

This would need special handling in some places which may not be likeable. I'd say we can retain `nvmf` too and back-compat can be handled, e.g. new control plane talking to old dataplane will see `nvmf` uri and infer tcp behaviour?

Reply:
> When RDMA is enabled, the uri will be nvmf+rdma (tcp will be implied).

Seems a bit confusing to imply tcp when using rdma? 
Whilst today tcp is used as fallback, this may not always be the case? What if we also use fc at some point, how would it work then?

Would it not be clearer to use +rdma+tcp for example? Or expose a list of URIs

Reply:
> Seems a bit confusing to imply tcp when using rdma?

I think it's a fair internal assumption and keeps things simpler to handle.

> Whilst today tcp is used as fallback, this may not always be the case? 

I'd think this behaviour will remain so always. Even if it doesn't, we may just need to change caller side to not fallback. Even if(at all) fc comes in, same fallback behaviour will remain. fc and rdma interconnect can't be together anyway.
I earlier thought about uri list too, but that changes the API path where we just send a string (single uri).

Reply:
Just a thought, since we are having this discussion, would it be more clear if this becomes a query param, like `?transport=tcp+rdma` that way the scheme remains same and we don't have to think about backward comp. ?

Reply:
As long as nvmf is kept when rdma is disabled, then I think backward compat is solved. Given that you'd need new control-plane to support rdma anyway.
Also to consider, if we want to support fc in the future then query parameter won't work because the traddr is different.
I think going forward, a list of URI's probably would be the saner way where you provide a list of potential connections. Also this would work in case we want to specify multiple networks for multipath for example.
But for now rdma+tcp probably is fine, and we can add list later.

Reply:
The nvme bootspec defines nvme URLs; might want to consult that.

Reply:
Thanks @gila .
We seem to be aligned with the guidance in spec that says uri scheme is to be `nvme+<trtype>`. We just slightly diverge and intend to use `nvmf+<trtype>+<trtype>` (i.e. `nvmf+rdma+tcp` to explicitly indicate support of both protocols) while still remaining compliant to RFC 3986.